core/include/initcall.h

```diff
@@ -8,15 +8,30 @@
 
 #include <scattered_array.h>
 #include <tee_api_types.h>
+#include <trace.h>
 
-typedef TEE_Result (*initcall_t)(void);
-
-#define __define_initcall(level, fn) \
-	SCATTERED_ARRAY_DEFINE_PG_ITEM_ORDERED(initcall, level, initcall_t) = \
-		(fn)
+struct initcall {
+	TEE_Result (*func)(void);
+#if TRACE_LEVEL >= TRACE_DEBUG
+	int level;
+	const char *func_name;
+#endif
+};
+
+#if TRACE_LEVEL >= TRACE_DEBUG
+#define __define_initcall(lvl, fn) \
+	SCATTERED_ARRAY_DEFINE_PG_ITEM_ORDERED(initcall, lvl, \
+					       struct initcall) = \
+		{ .func = (fn), .level = (lvl), .func_name = #fn, }
+#else
+#define __define_initcall(lvl, fn) \
+	SCATTERED_ARRAY_DEFINE_PG_ITEM_ORDERED(initcall, lvl, \
+					       struct initcall) = \
+		{ .func = (fn), }
```

Comment:
missing `.level = (lvl),`

Reply:
No, `.level` doesn't exist in the part of the #ifdef

Reply:
sorry, you're right.